File: example_py/test_aidk_others.py
"""
Test script for flexivaidk.so
Author: Flexiv
"""

import argparse
import json
import logging
import os
import sys
import time

logging.basicConfig(level=logging.INFO)

# Import Flexiv AIDK Python library
sys.path.insert(0, "../lib_py/")
import flexivaidk as aidk


def main(
    project,
    ip="127.0.0.1",
    detect_timeout=10,
    version="v3.1.0",
):
    client = aidk.AIDKClient(ip, detect_timeout)

    # AI state check
    while not client.is_ready():
        time.sleep(0.5)
    ai_status = client.get_current_state()
    logging.info("Current state code: {}".format(ai_status.status_code))
    logging.info("Current state name: {}".format(ai_status.status_name))
    logging.info("Current state message: {}".format(ai_status.status_message))

    if version >= "v3.1.0":
        """test variable setting."""
        # set variable
        logging.info(
            "Setable Variables: {}".format(client.get_direct_setting_variables())
        )
        res = client.set_direct_setting_variables(
            {
                "extern_scalar": 3.1415926,
                "extern_text": "world",
                "bool_check": True,
                "extern_int": 2,
            }
        )
        logging.info(res.error_code)
        logging.info(res.error_msg)
        logging.info(
            "Setable Variables: {}".format(client.get_direct_setting_variables())
        )

        # bad case tests
        res = client.set_direct_setting_variables({"extern_scalar": "abc"})
        logging.info(res.error_code)
        logging.info(res.error_msg)

        res = client.set_direct_setting_variables({"extern_invalid": 3.1415926})
        logging.info(res.error_code)
        logging.info(res.error_msg)

        res = client.set_direct_setting_variables({"extern_scalar": True})
        logging.info(res.error_code)
        logging.info(res.error_msg)

        res = client.set_direct_setting_variables({"extern_shape": 1})
        logging.info(res.error_code)
        logging.info(res.error_msg)

    """ test config related """
    if version >= "v2.11.1":
        state_code = client.save_configs()
        if not state_code:
            logging.error("save config failed!")

        state_code = client.reload_configs(project)
        if not state_code:
            logging.error("reload config failed for project: %s", project)
        logging.info("reload project: %s, state: %s", project, state_code)

        state_code = client.warmup()
        if not state_code:
            logging.error("warmup failed!")

    """test file transfer."""
    if version >= "v3.1.0":
        # get run time info
        working_dir, program_path, params = client.get_runtime_info()
        logging.info("working dir: {}".format(working_dir))
        logging.info("program path: {}".format(program_path))
        logging.info("support directory: {}".format(params))

        # get files and folders
        files, folders = client.list_remote_files("model/")
        logging.info("files: {}".format(files))
        logging.info("folders: {}".format(folders))

        # get file info
        mtime, size = client.get_file_info("model/upload.sh")
        logging.info("modify time: {}, size: {}".format(mtime, size))

        # file transfer
        client.send_file("./test_grasping_with_rdk.py", "model/test.py")
        client.receive_file("model/test.py", "test.py")
        client.remove_path("model/test.py")

        client.send_folder("../config", "model/FILE_TRANSFER")
        client.receive_folder("model/FILE_TRANSFER", "./FILE_TRANSFER")
        client.remove_path("model/FILE_TRANSFER")

        files, folders = client.list_remote_files("model/")
        logging.info("files: {}".format(files))
        logging.info("folders: {}".format(folders))


if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="Test Noema")
    parser.add_argument(
        "--config",
        help="config file path",
        default="config/TUTORIAL.json",
        type=str,
    )
    parser.add_argument(
        "--detect-timeout",
        help="Timeout of detect command request.",
        type=float,
        default=10,
    )
    parser.add_argument("--ip", help="Ip in a.b.c.d.", type=str, default="localhost")
    parser.add_argument(
        "--version", help="NoemaApp version in vx.x.x", type=str, default="v2.11.0"
    )

    args = parser.parse_args()

    assert os.path.exists(args.config), "Test config file %s not exist!" % args.config

    with open(args.config, "r") as f:
        config = json.load(f)

    logging.info(str(args))
    logging.info("Test config:\n%s", config)

    main(
        config["project"],
        args.ip,
        args.detect_timeout,
        args.version,
    )

File: example/test_aidk_others.cpp
/**
 * @example test_aidk.cpp
 * @brief test of AIDK client
 *
 * @copyright Copyright (C) 2023 Flexiv Ltd. All Rights Reserved.
 */

#include "flexiv/ai/aidk.hpp"
#include <fstream>
#include <nlohmann/json.hpp>
#include <typeinfo>
using json = nlohmann::json;

bool load_config(json &js, std::string file_path,
                 std::vector<double> &camera_pose,
                 std::vector<double> &tcp_pose, std::vector<double> &tcp_force)
{
    std::cout << "Config File: " << file_path << std::endl;
    std::ifstream file(file_path);
    file >> js;
    camera_pose = {0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0};
    tcp_pose = {0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0};
    tcp_force = {0.0, 0.0, 0.0, 0.0, 0.0, 0.0};
    if (js["command"]["camera_pose"].type() != nlohmann::json::value_t::null) {
        camera_pose = js["command"]["camera_pose"].get<std::vector<double>>();
    }
    if (js["command"]["tcp_pose"].type() != nlohmann::json::value_t::null) {
        tcp_pose = js["command"]["tcp_pose"].get<std::vector<double>>();
    }
    if (js["command"]["tcp_force"].type() != nlohmann::json::value_t::null) {
        tcp_force = js["command"]["tcp_force"].get<std::vector<double>>();
    }
    return true;
}

int main(int argc, char **argv)
{
    // init AIDK
    flexiv::ai::AIDKClient client(argv[1], 10);

    // load config file
    json js;
    std::vector<double> camera_pose;
    std::vector<double> tcp_pose;
    std::vector<double> tcp_force;
    load_config(js, argv[2], camera_pose, tcp_pose, tcp_force);

    // AI state check
    while (!client.is_ready()) {
        std::this_thread::sleep_for(std::chrono::milliseconds(500));
    }
    flexiv::ai::AIStatus ai_status = client.get_current_state();
    std::cout << "current state code: " << ai_status.status_code << std::endl;
    std::cout << "current state name: " << ai_status.status_name << std::endl;
    std::cout << "current state message: " << ai_status.status_message
              << std::endl;
    std::cout << std::endl;

    std::string version = argv[3];

    // set variable
    if (version >= "v3.1.0") {
        std::unordered_map<std::string, flexiv::ai::value_variant> vars =
            client.get_direct_setting_variables();
        std::cout << "Setable Variables: " << std::endl;
        for (auto &pair : vars) {
            std::cout << pair.first << ": ";
            if (pair.second.is<int>()) {
                std::cout << pair.second.get<int>() << std::endl;
            } else if (pair.second.is<double>()) {
                std::cout << pair.second.get<double>() << std::endl;
            } else if (pair.second.is<float>()) {
                std::cout << pair.second.get<float>() << std::endl;
            } else if (pair.second.is<bool>()) {
                std::cout << pair.second.get<bool>() << std::endl;
            } else if (pair.second.is<std::string>()) {
                std::cout << pair.second.get<std::string>() << std::endl;
            }
        }
        std::cout << std::endl;

        std::unordered_map<std::string, flexiv::ai::value_variant> set_vars;
        set_vars["extern_scalar"].set<float>(3.1415926f);
        set_vars["extern_text"].set<std::string>("world");
        set_vars["bool_check"].set<bool>(true);
        set_vars["extern_int"].set<int>(2);
        client.set_direct_setting_variables(set_vars);

        set_vars["extern_scalar"].set<double>(3.141592653589793);
        set_vars["extern_text"].set<std::string>("world");
        set_vars["bool_check"].set<bool>(true);
        set_vars["extern_int"].set<int>(2);
        client.set_direct_setting_variables(set_vars);

        vars = client.get_direct_setting_variables();
        std::cout << "Setable Variables: " << std::endl;
        for (auto &pair : vars) {
            std::cout << pair.first << ": ";
            if (pair.second.is<int>()) {
                std::cout << pair.second.get<int>() << std::endl;
            } else if (pair.second.is<double>()) {
                std::cout << pair.second.get<double>() << std::endl;
            } else if (pair.second.is<float>()) {
                std::cout << pair.second.get<float>() << std::endl;
            } else if (pair.second.is<bool>()) {
                std::cout << pair.second.get<bool>() << std::endl;
            } else if (pair.second.is<std::string>()) {
                std::cout << pair.second.get<std::string>() << std::endl;
            }
        }
        std::cout << std::endl;

        // bad case tests
        flexiv::ai::Response res;
        set_vars.clear();
        set_vars["extern_scalar"].set<std::string>("abc");
        res = client.set_direct_setting_variables(set_vars);
        std::cout << "error code: " << res.error_code << std::endl;
        std::cout << "error message: " << res.error_msg << std::endl;

        set_vars.clear();
        set_vars["extern_invalid"].set<float>(3.1415926);
        res = client.set_direct_setting_variables(set_vars);
        std::cout << "error code: " << res.error_code << std::endl;
        std::cout << "error message: " << res.error_msg << std::endl;

        set_vars.clear();
        set_vars["extern_scalar"].set<bool>(true);
        res = client.set_direct_setting_variables(set_vars);
        std::cout << "error code: " << res.error_code << std::endl;
        std::cout << "error message: " << res.error_msg << std::endl;

        set_vars.clear();
        set_vars["extern_shape"].set<int>(1);
        res = client.set_direct_setting_variables(set_vars);
        std::cout << "error code: " << res.error_code << std::endl;
        std::cout << "error message: " << res.error_msg << std::endl;
    }

    // reload and warmup
    if (version >= "v2.11.1") {
        bool status = client.save_configs();
        if (!status)
            std::cerr << "save config failed!" << std::endl;

        status = client.reload_configs(js["project"]);
        if (!status)
            std::cerr << "reload config failed for project: " << js["project"]
                      << std::endl;
        std::cout << "reload project: " << js["project"] << " state: " << status
                  << std::endl;

        status = client.warmup();
        if (!status)
            std::cerr << "warmup failed!" << std::endl;
    }

    // File transfer related
    if (version >= "v3.1.0") {
        // get AI run time info
        auto [working_dir, program_path, params] = client.get_runtime_info();
        std::cout << "working dir: " << working_dir << std::endl;
        std::cout << "program path: " << program_path << std::endl;
        std::cout << "suppot dir: ";
        for (auto param : params) {
            std::cout << param << " ";
        }
        std::cout << std::endl;

        // get files and folders
        std::vector<std::string> files, folders;
        std::tie(files, folders) = client.list_remote_files("model/");
        std::cout << "files: ";
        for (auto file : files) {
            std::cout << file << " ";
        }
        std::cout << std::endl;
        std::cout << "folders: ";
        for (auto folder : folders) {
            std::cout << folder << " ";
        }
        std::cout << std::endl;

        // get file info
        auto [mtime, size] = client.get_file_info("model/upload.sh");
        std::cout << "mtime: " << mtime << " size: " << size << std::endl;

        // file transfer
        client.send_file("../include/flexiv/ai/aidk.hpp", "model/test.hpp");
        client.receive_file("model/test.hpp", "test.hpp");
        client.remove_path("model/test.hpp");

        client.send_folder("../config", "model/FILE_TRANSFER");
        client.receive_folder("model/FILE_TRANSFER", "./FILE_TRANSFER");
        client.remove_path("model/FILE_TRANSFER");

        std::tie(files, folders) = client.list_remote_files("model/");
        std::cout << "files: ";
        for (auto file : files) {
            std::cout << file << " ";
        }
        std::cout << std::endl;
        std::cout << "folders: ";
        for (auto folder : folders) {
            std::cout << folder << " ";
        }
        std::cout << std::endl;
    }

    return 0;
}
